app/models/institution_builder.rb

```diff
@@ -0,0 +1,416 @@
+
+    version
+  end
+
+  def self.initialize_with_weams(version_number)
+    columns = Weam::USE_COLUMNS.map(&:to_s)
+
+    institutions = Weam.select(columns).where(approved: true).map(&:attributes).each_with_object([]) do |weam, a|
+      a << Institution.new(weam.except('id').merge(version: version_number))
+    end
+
+    # No validations here, Weam was validated on import - don't need to waste time with unique checks
+    Institution.import institutions, validate: false, ignore: true
+  end
+
+  def self.add_crosswalk(version_number)
+    columns = Crosswalk::USE_COLUMNS.map(&:to_s).map { |col| %("#{col}" = crosswalks.#{col}) }.join(', ')
+
+    str = <<-SQL
+      UPDATE institutions SET #{columns}
+      FROM crosswalks
+      WHERE institutions.facility_code = crosswalks.facility_code
+        AND institutions.version = #{version_number};
+    SQL
+
+    Institution.connection.update(str)
+  end
+
+  def self.add_sva(version_number)
+    str = <<-SQL
+      UPDATE institutions SET
+        student_veteran = TRUE, student_veteran_link = svas.student_veteran_link
+      FROM svas
+      WHERE institutions.cross = svas.cross AND svas.cross IS NOT NULL
+        AND institutions.version = #{version_number};
+    SQL
+
+    Institution.connection.update(str)
+  end
+
+  def self.add_vsoc(version_number)
+    columns = Vsoc::USE_COLUMNS.map(&:to_s).map { |col| %("#{col}" = vsocs.#{col}) }.join(', ')
+
+    str = <<-SQL
+      UPDATE institutions SET #{columns}
+      FROM vsocs
+      WHERE institutions.facility_code = vsocs.facility_code
+        AND institutions.version = #{version_number};
+    SQL
+
+    Institution.connection.update(str)
+  end
+
+  def self.add_eight_key(version_number)
+    str = <<-SQL
+      UPDATE institutions SET eight_keys = TRUE
+      FROM eight_keys
+      WHERE institutions.cross = eight_keys.cross
+        AND eight_keys.cross IS NOT NULL
+        AND institutions.version = #{version_number};
+    SQL
+
+    Institution.connection.update(str)
+  end
+
+  def self.add_accreditation(version_number)
+    # Set the accreditation_type according to the hierarchy hybrid < national < regional
+    # We include only those accreditation that are institutional and currently active.
+    str = <<-SQL
+      UPDATE institutions SET accreditation_type = CASE
+        WHEN at_types @> '{ regional }' THEN 'regional'
+        WHEN at_types @> '{ national }' THEN 'national'
+        WHEN at_types @> '{ hybrid }' THEN 'hybrid'
+        ELSE NULL
+      END
+      FROM (
+        SELECT "cross", array_agg(DISTINCT(accreditation_type)) AS at_types
+        FROM accreditations
+        WHERE "cross" IS NOT NULL
+          AND accreditation_type IS NOT NULL
+          AND periods LIKE '%current%'
+          AND csv_accreditation_type = 'institutional'
+        GROUP BY "cross") AS cross_type_arr
+      WHERE cross_type_arr.cross = institutions.cross
+        AND institutions.version = #{version_number};
+    SQL
```

Comment:
This query seems dangerous at first glance. I will want us to make sure all of the columns in questions have appropriate indexes. Is there anyway to optimize these queries? Why do we need the CASE When? can that not be done with ruby ahead of time?